File: DihhJ Backend/routes/bitch.py
from fastapi import APIRouter, HTTPException, status, Query
from typing import Optional
from datetime import datetime
from bson import ObjectId
from schemas.tea_schemas import (
    BitchCreate, BitchUpdate, BitchOut, BitchResponse, BitchListResponse
)
from database import bitch_collection, user_collection, tea_collection, bitch_votes_collection
import pymongo

router = APIRouter(prefix="/bitch", tags=["Bitch Comments"])

# Helper function to convert MongoDB document to BitchOut
def bitch_doc_to_out(doc: dict) -> BitchOut:
    return BitchOut(
        id=str(doc["_id"]),
        content=doc["content"],
        author_id=str(doc["author_id"]),
        author_username=doc["author_username"],
        tea_id=str(doc["tea_id"]),
        parent_id=str(doc["parent_id"]) if doc.get("parent_id") else None,
        upvotes=doc.get("upvotes", 0),
        downvotes=doc.get("downvotes", 0),
        score=doc.get("upvotes", 0) - doc.get("downvotes", 0),
        depth=doc.get("depth", 0),
        reply_count=doc.get("reply_count", 0),
        created_at=doc["created_at"],
        updated_at=doc["updated_at"],
        replies=None  # Will be populated separately if needed
    )

# Helper function to get user by username (simple auth check)
async def get_current_user(username: str):
    user = await user_collection.find_one({"username": username})
    if not user:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="User not found. Please login first."
        )
    return user

@router.post("/{tea_id}/create", response_model=BitchResponse)
async def create_bitch(
    tea_id: str,
    bitch: BitchCreate, 
    username: str = Query(..., description="Username of the comment author")
):
    """
    Create a new bitch comment on a tea post.
    
    - **tea_id**: ID of the tea post to comment on
    - **content**: Content of the comment
    - **username**: Username of the commenter (required for authentication)
    """
    # Verify user exists
    user = await get_current_user(username)
    
    # Verify tea exists
    try:
        tea = await tea_collection.find_one({"_id": ObjectId(tea_id)})
    except:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid tea ID format"
        )
    
    if not tea:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Tea not found"
        )
    
    # Handle parent comment for threading
    parent_depth = 0
    if bitch.parent_id:
        try:
            parent_comment = await bitch_collection.find_one({"_id": ObjectId(bitch.parent_id)})
            if not parent_comment:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="Parent comment not found"
                )
            if str(parent_comment["tea_id"]) != tea_id:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Parent comment is not from the same tea post"
                )
            parent_depth = parent_comment.get("depth", 0)
        except Exception as e:
            if isinstance(e, HTTPException):
                raise e
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid parent comment ID format"
            )

    # Create bitch document
    bitch_doc = {
        "content": bitch.content,
        "author_id": user["_id"],
        "author_username": user["username"],
        "tea_id": ObjectId(tea_id),
        "parent_id": ObjectId(bitch.parent_id) if bitch.parent_id else None,
        "depth": parent_depth + 1 if bitch.parent_id else 0,
        "upvotes": 0,
        "downvotes": 0,
        "reply_count": 0,
        "created_at": datetime.utcnow(),
        "updated_at": datetime.utcnow()
    }
    
    # Insert bitch into database
    result = await bitch_collection.insert_one(bitch_doc)

    # Update parent comment reply count if this is a reply
    if bitch.parent_id:
        await bitch_collection.update_one(
            {"_id": ObjectId(bitch.parent_id)},
            {"$inc": {"reply_count": 1}}
        )

    # Update tea comment count
    await tea_collection.update_one(
        {"_id": ObjectId(tea_id)},
        {"$inc": {"comment_count": 1}}
    )

    # Get the created bitch
    created_bitch = await bitch_collection.find_one({"_id": result.inserted_id})

    return BitchResponse(
        message="Bitch comment created successfully!",
        bitch=bitch_doc_to_out(created_bitch)
    )

@router.get("/{tea_id}/list", response_model=BitchListResponse)
async def list_bitches(
    tea_id: str,
    skip: int = Query(0, ge=0, description="Number of comments to skip"),
    limit: int = Query(20, ge=1, le=100, description="Number of comments to return"),
    order: str = Query("desc", pattern="^(asc|desc)$", description="Sort order by creation time: asc or desc")
):
    """
    Get all bitch comments for a specific tea post with pagination.
    """
    # Verify tea exists
    try:
        tea = await tea_collection.find_one({"_id": ObjectId(tea_id)})
    except:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid tea ID format"
        )
    
    if not tea:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Tea not found"
        )
    
    # Determine sort order
    sort_order = pymongo.ASCENDING if order == "asc" else pymongo.DESCENDING
    
    # Get total count for this tea
    total = await bitch_collection.count_documents({"tea_id": ObjectId(tea_id)})
    
    # Get bitches with pagination and sorting
    cursor = bitch_collection.find({"tea_id": ObjectId(tea_id)}).sort("created_at", sort_order).skip(skip).limit(limit)
    bitches = await cursor.to_list(length=limit)
    
    # Convert to BitchOut objects
    bitch_list = [bitch_doc_to_out(bitch) for bitch in bitches]
    
    return BitchListResponse(
        message=f"Retrieved {len(bitch_list)} bitch comments for tea",
        bitches=bitch_list,
        total=total
    )

@router.get("/comment/{bitch_id}", response_model=BitchResponse)
async def get_bitch(bitch_id: str):
    """
    Get a specific bitch comment by ID.
    """
    try:
        bitch = await bitch_collection.find_one({"_id": ObjectId(bitch_id)})
    except:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid bitch comment ID format"
        )
    
    if not bitch:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Bitch comment not found"
        )
    
    return BitchResponse(
        message="Bitch comment retrieved successfully",
        bitch=bitch_doc_to_out(bitch)
    )

@router.put("/comment/{bitch_id}", response_model=BitchResponse)
async def update_bitch(
    bitch_id: str, 
    bitch_update: BitchUpdate, 
    username: str = Query(..., description="Username of the comment owner")
):
    """
    Update a bitch comment. Only the author can update their comment.
    """
    # Verify user exists
    user = await get_current_user(username)
    
    try:
        # Check if bitch exists and user is the author
        bitch = await bitch_collection.find_one({"_id": ObjectId(bitch_id)})
    except:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid bitch comment ID format"
        )
    
    if not bitch:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Bitch comment not found"
        )
    
    if bitch["author_id"] != user["_id"]:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You can only update your own bitch comments"
        )
    
    # Update bitch
    await bitch_collection.update_one(
        {"_id": ObjectId(bitch_id)},
        {"$set": {"content": bitch_update.content, "updated_at": datetime.utcnow()}}
    )
    
    # Get updated bitch
    updated_bitch = await bitch_collection.find_one({"_id": ObjectId(bitch_id)})
    
    return BitchResponse(
        message="Bitch comment updated successfully!",
        bitch=bitch_doc_to_out(updated_bitch)
    )

@router.delete("/comment/{bitch_id}")
async def delete_bitch(bitch_id: str, username: str = Query(..., description="Username of the comment owner")):
    """
    Delete a bitch comment. Only the author can delete their comment.
    """
    # Verify user exists
    user = await get_current_user(username)
    
    try:
        # Check if bitch exists and user is the author
        bitch = await bitch_collection.find_one({"_id": ObjectId(bitch_id)})
    except:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid bitch comment ID format"
        )
    
    if not bitch:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Bitch comment not found"
        )
    
    if bitch["author_id"] != user["_id"]:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You can only delete your own bitch comments"
        )
    
    # Delete bitch
    await bitch_collection.delete_one({"_id": ObjectId(bitch_id)})
    
    return {"message": "Bitch comment deleted successfully!"}

@router.get("/user/{username}", response_model=BitchListResponse)
async def get_user_bitches(
    username: str,
    skip: int = Query(0, ge=0, description="Number of comments to skip"),
    limit: int = Query(20, ge=1, le=100, description="Number of comments to return")
):
    """
    Get all bitch comments by a specific user.
    """
    # Verify user exists
    user = await user_collection.find_one({"username": username})
    if not user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found"
        )
    
    # Get total count for this user
    total = await bitch_collection.count_documents({"author_id": user["_id"]})
    
    # Get bitches with pagination
    cursor = bitch_collection.find({"author_id": user["_id"]}).sort("created_at", pymongo.DESCENDING).skip(skip).limit(limit)
    bitches = await cursor.to_list(length=limit)
    
    # Convert to BitchOut objects
    bitch_list = [bitch_doc_to_out(bitch) for bitch in bitches]
    
    return BitchListResponse(
        message=f"Retrieved {len(bitch_list)} bitch comments by {username}",
        bitches=bitch_list,
        total=total
    )

@router.post("/comment/{bitch_id}/vote")
async def vote_bitch(
    bitch_id: str,
    vote_type: str = Query(..., pattern="^(upvote|downvote)$", description="Vote type: upvote or downvote"),
    username: str = Query(..., description="Username of the voter")
):
    """
    Vote on a bitch comment (upvote or downvote).
    Users can change their vote or remove it by voting the same way again.
    """
    # Verify user exists
    user = await get_current_user(username)

    try:
        # Check if bitch exists
        bitch = await bitch_collection.find_one({"_id": ObjectId(bitch_id)})
    except:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid bitch comment ID format"
        )

    if not bitch:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Bitch comment not found"
        )

    # Check if user already voted on this comment
    existing_vote = await bitch_votes_collection.find_one({
        "bitch_id": ObjectId(bitch_id),
        "user_id": user["_id"]
    })

    if existing_vote:
        # User has already voted - don't allow multiple votes
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"You have already voted on this comment with a {existing_vote['vote_type']}"
        )
    else:
        # Create new vote
        vote_doc = {
            "bitch_id": ObjectId(bitch_id),
            "user_id": user["_id"],
            "vote_type": vote_type,
            "created_at": datetime.utcnow(),
            "updated_at": datetime.utcnow()
        }

        await bitch_votes_collection.insert_one(vote_doc)

        # Update bitch vote counts
        if vote_type == "upvote":
            await bitch_collection.update_one(
                {"_id": ObjectId(bitch_id)},
                {"$inc": {"upvotes": 1}}
            )
        else:
            await bitch_collection.update_one(
                {"_id": ObjectId(bitch_id)},
                {"$inc": {"downvotes": 1}}
            )

        return {"message": f"{vote_type.capitalize()} added successfully!"}

@router.get("/comment/{bitch_id}/user-vote")
async def get_user_vote_status(
    bitch_id: str,
    username: str = Query(..., description="Username to check vote status for")
):
    """
    Get the current user's vote status for a comment.
    Returns the user's vote type or null if they haven't voted.
    """
    # Verify user exists
    user = await get_current_user(username)

    try:
        # Check if comment exists
        bitch = await bitch_collection.find_one({"_id": ObjectId(bitch_id)})
    except:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid bitch comment ID format"
        )

    if not bitch:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Bitch comment not found"
        )

    # Check if user has voted on this comment
    existing_vote = await bitch_votes_collection.find_one({
        "bitch_id": ObjectId(bitch_id),
        "user_id": user["_id"]
    })

    return {
        "bitch_id": bitch_id,
        "user_vote": existing_vote["vote_type"] if existing_vote else None,
        "has_voted": existing_vote is not None
    }

File: DihhJ Backend/routes/bitch.js
const express = require('express');
const { ObjectId } = require('mongodb');
const { getDB } = require('../config/database');

const router = express.Router();

// Get comments for a tea post
router.get('/:teaId/list', async (req, res) => {
  try {
    const { teaId } = req.params;
    const { skip = 0, limit = 20, sort_by = 'new', order = 'desc' } = req.query;
    
    if (!ObjectId.isValid(teaId)) {
      return res.status(400).json({
        error: 'Invalid tea ID',
        message: 'Please provide a valid tea ID'
      });
    }
    
    const db = getDB();
    const commentsCollection = db.collection('comments');
    
    // Build sort
    let sortOptions = {};
    if (sort_by === 'hot') {
      sortOptions = { score: order === 'desc' ? -1 : 1 };
    } else {
      sortOptions = { created_at: order === 'desc' ? -1 : 1 };
    }
    
    // Get comments
    const comments = await commentsCollection
      .find({ tea_id: teaId })
      .sort(sortOptions)
      .skip(parseInt(skip))
      .limit(parseInt(limit))
      .toArray();
    
    // Get total count
    const total = await commentsCollection.countDocuments({ tea_id: teaId });
    
    res.json({
      comments: comments.map(comment => ({
        id: comment._id,
        content: comment.content,
        author: comment.author,
        tea_id: comment.tea_id,
        upvotes: comment.upvotes || 0,
        downvotes: comment.downvotes || 0,
        score: comment.score || 0,
        created_at: comment.created_at,
        updated_at: comment.updated_at
      })),
      total,
      skip: parseInt(skip),
      limit: parseInt(limit),
      hasMore: (parseInt(skip) + parseInt(limit)) < total
    });
    
  } catch (error) {
    console.error('Get comments error:', error);
    res.status(500).json({
      error: 'Failed to fetch comments',
      message: 'An error occurred while fetching comments'
    });
  }
});

// Create new comment
router.post('/:teaId/create', async (req, res) => {
  try {
    const { teaId } = req.params;
    const { content } = req.body;
    const { username } = req.query;
    
    // Validation
    if (!username) {
      return res.status(400).json({
        error: 'Missing username',
        message: 'Username is required'
      });
    }
    
    if (!ObjectId.isValid(teaId)) {
      return res.status(400).json({
        error: 'Invalid tea ID',
        message: 'Please provide a valid tea ID'
      });
    }
    
    if (!content || content.trim().length === 0) {
      return res.status(400).json({
        error: 'Missing content',
        message: 'Comment content is required'
      });
    }
    
    if (content.length < 1 || content.length > 1000) {
      return res.status(400).json({
        error: 'Invalid content',
        message: 'Comment must be between 1 and 1000 characters'
      });
    }
    
    const db = getDB();
    const teasCollection = db.collection('teas');
    const commentsCollection = db.collection('comments');
    
    // Check if tea exists
    const tea = await teasCollection.findOne({ _id: new ObjectId(teaId) });
    if (!tea) {
      return res.status(404).json({
        error: 'Tea not found',
        message: 'Tea post does not exist'
      });
    }
    
    // Create comment
    const newComment = {
      content: content.trim(),
      author: username,
      tea_id: teaId,
      upvotes: 0,
      downvotes: 0,
      score: 0,
      created_at: new Date(),
      updated_at: new Date()
    };
    
    const result = await commentsCollection.insertOne(newComment);
    
    res.status(201).json({
      message: 'Comment posted successfully!',
      comment: {
        id: result.insertedId,
        ...newComment
      }
    });
    
  } catch (error) {
    console.error('Create comment error:', error);
    res.status(500).json({
      error: 'Failed to create comment',
      message: 'An error occurred while creating the comment'
    });
  }
});

// Update comment
router.put('/comment/:commentId', async (req, res) => {
  try {
    const { commentId } = req.params;
    const { content } = req.body;
    const { username } = req.query;
    
    // Validation
    if (!username) {
      return res.status(400).json({
        error: 'Missing username',
        message: 'Username is required'
      });
    }
    
    if (!ObjectId.isValid(commentId)) {
      return res.status(400).json({
        error: 'Invalid comment ID',
        message: 'Please provide a valid comment ID'
      });
    }
    
    if (!content || content.trim().length === 0) {
      return res.status(400).json({
        error: 'Missing content',
        message: 'Comment content is required'
      });
    }
    
    if (content.length < 1 || content.length > 1000) {
      return res.status(400).json({
        error: 'Invalid content',
        message: 'Comment must be between 1 and 1000 characters'
      });
    }
    
    const db = getDB();
    const commentsCollection = db.collection('comments');
    
    // Check if comment exists and user owns it
    const comment = await commentsCollection.findOne({ _id: new ObjectId(commentId) });
    if (!comment) {
      return res.status(404).json({
        error: 'Comment not found',
        message: 'Comment does not exist'
      });
    }
    
    if (comment.author !== username) {
      return res.status(403).json({
        error: 'Unauthorized',
        message: 'You can only edit your own comments'
      });
    }
    
    // Update comment
    await commentsCollection.updateOne(
      { _id: new ObjectId(commentId) },
      {
        $set: {
          content: content.trim(),
          updated_at: new Date()
        }
      }
    );
    
    // Get updated comment
    const updatedComment = await commentsCollection.findOne({ _id: new ObjectId(commentId) });
    
    res.json({
      message: 'Comment updated successfully!',
      comment: {
        id: updatedComment._id,
        content: updatedComment.content,
        author: updatedComment.author,
        tea_id: updatedComment.tea_id,
        upvotes: updatedComment.upvotes || 0,
        downvotes: updatedComment.downvotes || 0,
        score: updatedComment.score || 0,
        created_at: updatedComment.created_at,
        updated_at: updatedComment.updated_at
      }
    });
    
  } catch (error) {
    console.error('Update comment error:', error);
    res.status(500).json({
      error: 'Failed to update comment',
      message: 'An error occurred while updating the comment'
    });
  }
});

// Delete comment
router.delete('/comment/:commentId', async (req, res) => {
  try {
    const { commentId } = req.params;
    const { username } = req.query;
    
    // Validation
    if (!username) {
      return res.status(400).json({
        error: 'Missing username',
        message: 'Username is required'
      });
    }
    
    if (!ObjectId.isValid(commentId)) {
      return res.status(400).json({
        error: 'Invalid comment ID',
        message: 'Please provide a valid comment ID'
      });
    }
    
    const db = getDB();
    const commentsCollection = db.collection('comments');
    const commentVotesCollection = db.collection('comment_votes');
    
    // Check if comment exists and user owns it
    const comment = await commentsCollection.findOne({ _id: new ObjectId(commentId) });
    if (!comment) {
      return res.status(404).json({
        error: 'Comment not found',
        message: 'Comment does not exist'
      });
    }
    
    if (comment.author !== username) {
      return res.status(403).json({
        error: 'Unauthorized',
        message: 'You can only delete your own comments'
      });
    }
    
    // Delete comment and its votes
    await commentsCollection.deleteOne({ _id: new ObjectId(commentId) });
    await commentVotesCollection.deleteMany({ comment_id: commentId });
    
    res.json({
      message: 'Comment deleted successfully!'
    });
    
  } catch (error) {
    console.error('Delete comment error:', error);
    res.status(500).json({
      error: 'Failed to delete comment',
      message: 'An error occurred while deleting the comment'
    });
  }
});

// Vote on comment
router.post('/comment/:commentId/vote', async (req, res) => {
  try {
    const { commentId } = req.params;
    const { vote_type } = req.query;
    const { username } = req.query;

    // Validation
    if (!username) {
      return res.status(400).json({
        error: 'Missing username',
        message: 'Username is required'
      });
    }

    if (!ObjectId.isValid(commentId)) {
      return res.status(400).json({
        error: 'Invalid comment ID',
        message: 'Please provide a valid comment ID'
      });
    }

    if (!vote_type || !['upvote', 'downvote'].includes(vote_type)) {
      return res.status(400).json({
        error: 'Invalid vote type',
        message: 'Vote type must be "upvote" or "downvote"'
      });
    }

    const db = getDB();
    const commentsCollection = db.collection('comments');
    const commentVotesCollection = db.collection('comment_votes');

    // Check if comment exists
    const comment = await commentsCollection.findOne({ _id: new ObjectId(commentId) });
    if (!comment) {
      return res.status(404).json({
        error: 'Comment not found',
        message: 'Comment does not exist'
      });
    }

    // Check for existing vote
    const existingVote = await commentVotesCollection.findOne({
      user_id: username,
      comment_id: commentId
    });

    let upvoteChange = 0;
    let downvoteChange = 0;

    if (existingVote) {
      // Remove previous vote
      if (existingVote.vote_type === 'upvote') {
        upvoteChange -= 1;
      } else {
        downvoteChange -= 1;
      }

      // If same vote type, remove vote entirely
      if (existingVote.vote_type === vote_type) {
        await commentVotesCollection.deleteOne({ _id: existingVote._id });
      } else {
        // Update to new vote type
        await commentVotesCollection.updateOne(
          { _id: existingVote._id },
          { $set: { vote_type, updated_at: new Date() } }
        );

        if (vote_type === 'upvote') {
          upvoteChange += 1;
        } else {
          downvoteChange += 1;
        }
      }
    } else {
      // Create new vote
      await commentVotesCollection.insertOne({
        user_id: username,
        comment_id: commentId,
        vote_type,
        created_at: new Date()
      });

      if (vote_type === 'upvote') {
        upvoteChange += 1;
      } else {
        downvoteChange += 1;
      }
    }

    // Update comment vote counts
    const newUpvotes = Math.max(0, (comment.upvotes || 0) + upvoteChange);
    const newDownvotes = Math.max(0, (comment.downvotes || 0) + downvoteChange);
    const newScore = newUpvotes - newDownvotes;

    await commentsCollection.updateOne(
      { _id: new ObjectId(commentId) },
      {
        $set: {
          upvotes: newUpvotes,
          downvotes: newDownvotes,
          score: newScore,
          updated_at: new Date()
        }
      }
    );

    // Get updated comment
    const updatedComment = await commentsCollection.findOne({ _id: new ObjectId(commentId) });

    res.json({
      message: 'Vote recorded!',
      comment: {
        id: updatedComment._id,
        content: updatedComment.content,
        author: updatedComment.author,
        tea_id: updatedComment.tea_id,
        upvotes: updatedComment.upvotes,
        downvotes: updatedComment.downvotes,
        score: updatedComment.score,
        created_at: updatedComment.created_at,
        updated_at: updatedComment.updated_at
      }
    });

  } catch (error) {
    console.error('Vote comment error:', error);
    res.status(500).json({
      error: 'Failed to vote',
      message: 'An error occurred while voting on the comment'
    });
  }
});

// Get user's vote status for a comment
router.get('/comment/:commentId/user-vote', async (req, res) => {
  try {
    const { commentId } = req.params;
    const { username } = req.query;

    if (!username) {
      return res.status(400).json({
        error: 'Missing username',
        message: 'Username is required'
      });
    }

    if (!ObjectId.isValid(commentId)) {
      return res.status(400).json({
        error: 'Invalid comment ID',
        message: 'Please provide a valid comment ID'
      });
    }

    const db = getDB();
    const commentVotesCollection = db.collection('comment_votes');

    const vote = await commentVotesCollection.findOne({
      user_id: username,
      comment_id: commentId
    });

    res.json({
      user_vote: vote ? vote.vote_type : null
    });

  } catch (error) {
    console.error('Get user comment vote error:', error);
    res.status(500).json({
      error: 'Failed to fetch vote status',
      message: 'An error occurred while fetching vote status'
    });
  }
});

// Get user's comments
router.get('/user/:username', async (req, res) => {
  try {
    const { username } = req.params;
    const { skip = 0, limit = 10 } = req.query;

    const db = getDB();
    const commentsCollection = db.collection('comments');

    // Get user's comments
    const comments = await commentsCollection
      .find({ author: username })
      .sort({ created_at: -1 })
      .skip(parseInt(skip))
      .limit(parseInt(limit))
      .toArray();

    // Get total count
    const total = await commentsCollection.countDocuments({ author: username });

    res.json({
      comments: comments.map(comment => ({
        id: comment._id,
        content: comment.content,
        author: comment.author,
        tea_id: comment.tea_id,
        upvotes: comment.upvotes || 0,
        downvotes: comment.downvotes || 0,
        score: comment.score || 0,
        created_at: comment.created_at,
        updated_at: comment.updated_at
      })),
      total,
      skip: parseInt(skip),
      limit: parseInt(limit),
      hasMore: (parseInt(skip) + parseInt(limit)) < total
    });

  } catch (error) {
    console.error('Get user comments error:', error);
    res.status(500).json({
      error: 'Failed to fetch user comments',
      message: 'An error occurred while fetching user comments'
    });
  }
});

module.exports = router;
